Site/SiteCookieModule.php

```diff
@@ -1,229 +1,202 @@
+     * Set salt.
+     *
+     * @param mixed $salt
+     */
+    public function setSalt($salt)
+    {
+        $this->salt = $salt;
+    }
+
+    /**
+     * Sets a cookie.
+     *
+     * @param string $name   the name of the cookie to set
+     * @param mixed  $value  the value of the cookie
+     * @param mixed  $expiry the expiry date as a UNIX timestamp or a
+     *                       string parsable by strtotime()
+     * @param string $path   the URL path this cookie is valid for
+     * @param string $domain the domain this cookie is valid for
+     */
+    public function setCookie(
+        $name,
+        $value,
+        $expiry = null,
+        $path = '/',
+        $domain = null
+    ) {
+        $name = $this->getPrefix() . '_' . $name;
+
+        if ($expiry === null) {
+            $expiry = strtotime('+90 days');
+        } elseif (is_string($expiry)) {
+            $expiry = strtotime($expiry);
+        }
+
+        $cookie_value = SwatString::signedSerialize($value, $this->salt);
+
+        // TODO: get subdomain from application when on a multi-instance site.
+        // if ($domain = null)
+        //	$domain =
+
+        setcookie($name, $cookie_value, ['expires' => $expiry, 'path' => $path]);
+        // setcookie($name, $cookie_value, ['expires' => $expiry, 'path' => $path, 'domain' => $domain]);
+    }
+
+    /**
+     * Remove a cookie.
+     *
+     * @param string $name   the name of the cookie to set
+     * @param string $path   the URL path this cookie is valid for
+     * @param string $domain the domain this cookie is valid for
+     */
+    public function removeCookie($name, $path = '/', $domain = null)
+    {
+        $name = $this->getPrefix() . '_' . $name;
+
+        // Set expiry time to the past. The expiry of 25 hours in the past is
+        // used because time() uses the server's local time and some browsers
+        // use local time rather than UTC to trigger cookie deletion. 25 hours
+        // takes into account all time-zone differences.
+        $expiry = time() - 3600 * 25;
+
+        // Some browsers set the cookie value to 'deleted' when an empty string
+        // is used as a cookie value. The value '0' is chosen instead for
+        // unsetting cookies.
+        $value = 0;
+
+        // TODO: get from application when on a multi-instance site.
+        // if ($domain = null)
+        //	$domain =
+
+        setcookie($name, $value, ['expires' => $expiry, 'path' => $path]);
+        // setcookie($name, $value, ['expires' => $expiry, 'path' => $path, 'domain' => $domain]);
+
+        unset($_COOKIE[$name]);
+    }
+
+    /**
+     * Gets the module features this module depends on.
+     *
+     * @return array an array of {@link SiteModuleDependency} objects defining
+     *               the features this module depends on
+     */
+    public function depends()
+    {
+        $depends = parent::depends();
+
+        if ($this->app->hasModule('SiteMultipleInstanceModule')) {
```

Comment:
this could use `SiteMultipleInstanceModule::class`

Reply:
From what I could tell, `addModule(SiteApplicationModule $class, string $id)` adds the modules to the array and builds up the `$modules_by_provides` array.  But it wasn't clear if `hasModule()` and `getModule()` expect the classname as parameters, or if that was just coincidental and we're using the (arbitrary) `$id` string from when the module was added.

It certainly _looks_ like in all instances we are using the class name and could make the switch you suggest.  I just wasn't sure.

Reply:
Good observation. It looks like the code is intended to use "features" for this param. In theory a module could provide multiple features, or multiple modules could provide the same feature, and you could check if the feature is supported.

Leaving as a string makes sense.